extensions/etcdsnapshot/etcdsnapshot.go

```diff
@@ -240,3 +248,117 @@ func RestoreRKE2K3SSnapshot(client *rancher.Client, clusterName string, snapshot
 
 	return nil
 }
+
+// RetentionLimitCheckRKE1 is a check that validates that the number of automatic snapshots on the cluster is under the retention limit
+func RetentionLimitCheckRKE1(client *rancher.Client, clusterName string) error {
+	clusterID, err := clusters.GetClusterIDByName(client, clusterName)
+	if err != nil {
+		return err
+	}
+
+	clusterResp, err := client.Management.Cluster.ByID(clusterID)
+	if err != nil {
+		return err
+	}
+
+	retentionLimit := clusterResp.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.Retention
+
+	var s3Enabled bool = false
+	s3Config := clusterResp.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.S3BackupConfig
+	if s3Config != nil {
+		s3Enabled = true
+	}
+
+	existingSnapshots, err := GetRKE1Snapshots(client, clusterName)
+	automaticSnapshots := []management.EtcdBackup{}
+	for _, snapshot := range existingSnapshots {
+		if !snapshot.Manual {
+			automaticSnapshots = append(automaticSnapshots, snapshot)
+		}
+	}
+
+	clusterNodes, err := nodes.GetNodes(client, clusterID, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	expectedSnapshotsNum := len(clusterNodes) * int(retentionLimit)
+	if s3Enabled {
+		expectedSnapshotsNum += int(retentionLimit)
+	}
+
+	if len(automaticSnapshots) > expectedSnapshotsNum {
+		msg := fmt.Sprintf("retention limit exceeded: expected %d snapshots, found %d snapshots", expectedSnapshotsNum, len(automaticSnapshots))
+		logrus.Info(automaticSnapshots)
+		return errors.New(msg)
+	}
+
+	logrus.Infof("Snapshot retention limit respected, Snapshots Expected: %v Snapshots Found: %v", expectedSnapshotsNum, len(automaticSnapshots))
+
+	return nil
+}
+
+// RetentionLimitCheckRKE2K3S is a check that validates that the number of automatic snapshots on the cluster is under the retention limit
+func RetentionLimitCheckRKE2K3S(client *rancher.Client, clusterName string) error {
+	v1ClusterID, err := clusters.GetV1ProvisioningClusterByName(client, clusterName)
+	if err != nil {
+		return err
+	}
+
+	clusterObj, err := client.Steve.SteveType(defaults.ProvisioningSteveResourceType).ByID(v1ClusterID)
+	if err != nil {
+		return err
+	}
+
+	retentionLimit := clusterObj.Spec.(map[string]interface{})["rkeConfig"].(map[string]interface{})["etcd"].(map[string]interface{})["snapshotRetention"].(float64)
```

Comment:
i think what you have could be okay, but typically we convert the spec to a steve/management object first, then pull the variable from it. i.e. this would probably be 
```
	spec := &provv1.ClusterSpec{}
	err = steveV1.ConvertToK8sType(clusterObj.Spec, spec)
	if errr...
```
instead of type casting through go

Reply:
Glad there is an easier way to do this, i'll change it.